Add support for multiple backends and add HTML backend

hreymond opened the pull request:
Add support for multiple backend and add an HTML backend to the compiler.

Add new options for the compiler, namely
```
  -f, --format <FORMAT>      Output format [default: tikz]
  -t, --template <TEMPLATE>  Template to use, if any
  -o, --output <FILE>        Output file. If not present, will output to stdout
̀```

hreymond committed:
Added different output formats

hreymond committed:
Updated README.md

Added info on Seri compiler usage

hreymond committed:
Added various params and HTML Backend

- Added HTML Backend
- Added format,  template  and output file parameters to seri compiler
- Added HTML template
- Improved compiler error handling (hopefully)

ororktn commented on README.md:
```diff
@@ -1,4 +1,42 @@
+
+TODO : Write a proper language specification
+
+## TODO
```
You can check the HTML > Minimal box, and precise what you would like to add to make it a "not so minimal" HTML output

hreymond replied:
Done !

hreymond committed:
Je déteste les espaces insécables

ororktn commented on Cargo.toml:
Could maybe also deserve a minor version bump I guess, it's satisfying :3

hreymond replied:
I love it :tada:

ororktn commented on src/main.rs:
```diff
@@ -22,24 +22,55 @@
 };
 use clap::Parser;
-use std::{fs, io::Read};
+use passes::html::{HTMLBackend, HTMLBackendOptions};
```
To factor things out, you may want to move this in the `use crate` directive (lines 22 and onwards)

Lymkwi commented on src/passes/html.rs:
```diff
@@ -0,0 +1,87 @@
+    }
+}
+
+fn event_to_string(event: &Event) -> String {
```
Why not directly implement this as an impl of [`ToString`](https://doc.rust-lang.org/std/string/trait.ToString.html) on `Event` ? You would have to also change line 79 to read `str.push(&event.to_string())` rather than `str.push(event_to_string(&event).as_str())`.

hreymond replied:
I like the idea ! However, for the data and the passes to stay separated, It would be nice to:

1. Be able to implement [ToString](https://doc.rust-lang.org/std/string/trait.ToString.html) on `Event` in `html.rs` (since only the HTML backend need to convert an Event to an HTML String)
2. Be able to *specialize* the `to_string` function (for example  `to_string<HTML>()` for a HTML representation, and `to_string<TIKZ>()`for a tikz representation), since every backend could need its own `to_string` implementation

Have you got an idea on how we could do this ? (I don't master Rust enough ^^)

Docteur-Lalla replied:
Indeed the to_string here serializes for the HTML backend and not for an "absolute context". Maybe we could have a `ToHTML` trait or something.

Docteur-Lalla replied:
Serde ?

Lymkwi replied:
Serde would also be sort of an "absolute context" converter I think, but the middle ground of a `ToHTML` trait is fair enough, I think it solves the need to convert in this module specifically without implying a universal format for `Event`.

hreymond replied:
Done with a ToHTML trait :wink:

Lymkwi commented on src/passes/html.rs:
```diff
@@ -0,0 +1,87 @@
+
+fn get_template() -> Result<String, HTMLBackendCompilationError> {
+    let options = OPTIONS.lock().unwrap();
+    if let Some(path) = &options.template_path {
```
You could write this entire section as a `map_or_else` instead of an if/let/else:
```
options.template_path.as_ref().map_or_else(|| Ok(include_str!("../../data/template.html").to_string()), |path| match std::fs::read_to_string(path) {
        Ok(template) => Ok(template),
        Err(err) => Err(HTMLBackendCompilationError::CouldNotReadTemplate(err)),
    })
```

Docteur-Lalla replied:
In the present case, wouldn't it be better to go for a `match` construct itself? All these closure make the code complicated, while it could be more straightforward:

```rust
match options.template_path.as_ref() {
    None => Ok(include_str!("../../data/template.html").to_string()),
    Some(path) => std::fs::read_to_string(path).map_err(HTMLBackendCompilationError::from),
}
```

Lymkwi replied:
Agreed, match makes it easier to read. I'm generally okay with a bunch of closures in my code but I agree that it's not for everyone.

hreymond replied:
I agree, this is a bit of a mess, I've modified this part of the code, it should be better !

Lymkwi commented on src/passes/html.rs:
```diff
@@ -0,0 +1,87 @@
+fn event_to_string(event: &Event) -> String {
+    let duration = std::cmp::max(event.duration / 30, 1);
+    let class = event.event_type.to_string();
+    let mut res = String::from(format!("\t\t<td class=\"{class}\" rowspan=\"{duration}\">"));
```
`format!` already outputs a `String`.

Lymkwi commented on src/passes/html.rs:
```diff
@@ -0,0 +1,87 @@
+    let class = event.event_type.to_string();
+    let mut res = String::from(format!("\t\t<td class=\"{class}\" rowspan=\"{duration}\">"));
+    let (title, _) = event.title.split_at(min(event.title.len(), 25));
+    res.push_str(format!("{}", title).as_str());
```
You don't need to `format!` and then `.as_str`, because `title` is already a `&str`.

Lymkwi commented on src/passes/html.rs:
```diff
@@ -0,0 +1,87 @@
+}
+
+/// Options of the HTML backend
+static OPTIONS: Mutex<HTMLBackendOptions> = Mutex::new(HTMLBackendOptions {
```
Genuinely curious, what is the use of the `Mutex` here? I don't recall there being any parallelism or concurrency in this whole project. I may be mistaken, however.

Docteur-Lalla replied:
I'd also argue that using a static option object make functions non reentrant and as such will make unit-testing even more difficult. We should just use a regular `HTMLBackendOptions` object that we pass as parameter to the functions that need it.

hreymond replied:
I wanted to be able to configure the backend pass without modifiyng the interface provided in `passes.rs`, and more importantly the signature of `apply`,  used in `chain_pass`: `fn apply(_: T) -> Result<Self::Residual, Self::Error>;`.

The only solution I found for now is to have a static variable that contains the parameters of the pass, that can be used afterwards by `apply`. However, having a static variable means having potential data races, which rust compiler does not like. Thus, i protected the variable with a mutex, which is really ugly I agree.

If you have any idea on how to handle parameters without changing `apply`signature, I'm in ! Otherwise we could think of changing the interface, but it's another whole story and we should create a new issue/branch/PR

Lymkwi replied:
I think it should not be too hard to spin up another method in the Trait that can be invoked with an argument of a type associated to the trait (with default to `()` if not given by the `impl`). I could do a quick demo and PR for it.

Lymkwi replied:
Quick and dirty proposal available at 12659d836bf599654b0eeeafba0fc0271981cb1b

Lymkwi commented on src/passes/html.rs:
```diff
@@ -0,0 +1,87 @@
+        let mut str = String::new();
+        str.push_str("<table><tbody>\n");
+        for i in 9..21 {
+            str.push_str(format!("\t<tr><th>{}:00</th>", i).as_str());
```
You can put the `i` in the brackets in the format string :smiley: 
```
"\t<tr><th>{i}:00</th>"
```

EDIT: Same comment for line 82.

Lymkwi commented on src/main.rs:
```diff
@@ -63,8 +105,27 @@ fn main() {
+    let output: Result<String, CompilerError> = match args.format.as_str() {
+        "tikz" => generate_tikz(&content).map_err(|err| CompilerError::CouldNotGenerateTikz(err)),
+        "html" => generate_html(html_opts, &content)
+            .map_err(|err| CompilerError::CouldNotGenerateHTML(err)),
```
No need to create a closure here, just pass `CompilerError::CouldNotGenerateHTML` as argument.

Lymkwi replied:
Oh and same for line 113!

hreymond replied:
I didn't realised, that's so much simpler oO
I just replaced the awful closure with `CompilerError::from`, that's way better

Lymkwi commented on src/main.rs:
```diff
@@ -63,8 +105,27 @@ fn main() {
+
+    match output {
+        Ok(data) => match args.output {
+            Some(file_path) => {
```
You could also use a `map_or_else` here, as suggested by `clippy`, but I agree that it could be a bit overkill. For reference, it would be:
```
        Ok(data) => args.output.map_or_else(|| println!("{data}"), |file_path| {
                 let mut file =
                     File::create(&file_path).expect(format!("Couldnt open {file_path}").as_str());
                 file.write_all(data.as_bytes())
                     .expect("Couldnt write to {file_path}");
             }),
```

Lymkwi commented on src/main.rs:
```diff
@@ -63,8 +105,27 @@ fn main() {
+        Ok(data) => match args.output {
+            Some(file_path) => {
+                let mut file =
+                    File::create(&file_path).expect(format!("Couldnt open {file_path}").as_str());
```
`.expect` is nice for constant strings, but if you are going to format a string, you might want to use an `unwrap_or_else` and voluntarily `panic!` with the formatting in it like `panic!("Couldn't open {file_path}")`.

hreymond replied:
Good idea ! Done :wink:

Lymkwi reviewed (changes requested):
Overall very good, just finnicky details here and there that could easily be improved :fox_face: :+1:

Docteur-Lalla commented on src/passes/html.rs:
```diff
@@ -0,0 +1,87 @@
+            str.push_str(format!("\t<tr><th>{}:00</th>", i).as_str());
+            for event in events
+                .iter()
+                .filter(|ev| ev.start_date.hour() >= i && ev.start_date.hour() < i + 1)
```
I don't get this `filter`. You want the hour to be greater or equal to `i` and less than `i + 1`. Given that `hour` returns a `u32` and `i` is an integer itself, isn't it an equality test?

hreymond replied:
For now it's only handling event that start at a round hour. In the future, I hope to add support for half-hour and minute precision.

Docteur-Lalla commented on src/passes/html.rs:
```diff
@@ -0,0 +1,87 @@
+    CouldNotReadTemplate(#[from] std::io::Error),
+}
+
+fn get_template() -> Result<String, HTMLBackendCompilationError> {
```
Given that this function can only fail while reading the template (in `read_to_string`), I would suggest to either return an `std::io::Error` or create a newtype over it. I'm not convince we need a bigger error type, especially since functions that call `get_template` can create a `HTMLBackendCompilationError` from an `std::io::Error` using `from`.

Docteur-Lalla commented on src/event.rs:
Please use `cargo fmt` to reformat the code according to *the standard:tm:*

hreymond replied:
This is `cargo fmt` that changed the `src/event.rs` format, I didn't opened the file. Do I need to configure something to get the same formatting as you ? (line size for example)

ororktn commented on src/main.rs:
```diff
@@ -22,24 +22,55 @@
     #[arg(help = "File to compile. If not present, will read from standard input")]
     file: Option<String>,
+    #[arg(short, long, value_name = "FORMAT", help = "Output format", default_value_t = String::from("tikz"))]
+    format: String,
```
Try using directly an enum, see the [ValueEnum](https://docs.rs/clap/latest/clap/trait.ValueEnum.html) trait and [this StackOverflow reply](https://stackoverflow.com/a/73184990)

hreymond replied:
Thanks for the link ! I didn't know it was possible

hreymond committed:
Simplified format and output handling + cleaning

hreymond reviewed (commented):
Thank you very much for your detailed reviews :heart: 
I commited a new version of the code to fix some problem mentioned in the reviews. For some comment, I'm looking for ideas, feel free to propose code or commit directly to the branch.
I'll add another commit to add the ToHTML trait, as proposed by @Docteur-Lalla and @Lugrim

Lymkwi reviewed (commented):
I think I would like to see the `ToHTML` trait added before merging, but otherwise, looks good, and all the little details have been fixed.

Lymkwi committed:
Implement compiling passes with context structures

Add an `apply_with` method to the `CompilingPass` trait such that pass
inputs can be called with `apply_with` and an object containing
associated context/values.

This commit is a proof of work to replace the use of a mutex and global
variable in #12 .

In order to modify that pull request to use it, please apply the patch
below.

Signed-off-by: Amélie "Lymkwi" Gonzalez <lymkwi@vulpinecitrus.info>

---

diff --git a/src/main.rs b/src/main.rs
index 2a569f1..7a34ec2 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -89,10 +89,9 @@ fn generate_html(
     options: HTMLBackendOptions,
     content: &str,
 ) -> Result<String, HTMLBackendCompilationError> {
-    HTMLBackend::configure(options);
     content
         .chain_pass::<ParseTimetable>()?
-        .chain_pass::<HTMLBackend>()
+        .chain_pass_with::<HTMLBackend, HTMLBackendOptions>(options)
 }

 /// Write the output to a file or to stdout
diff --git a/src/passes/html.rs b/src/passes/html.rs
index 415f885..14cc854 100644
--- a/src/passes/html.rs
+++ b/src/passes/html.rs
@@ -1,22 +1,27 @@
 //! HTML backend
 use crate::{event::Event, passes::CompilingPass};
 use chrono::Timelike;
-use std::{cmp::min, str::FromStr, sync::Mutex};
+use std::{cmp::min, str::FromStr};
 use thiserror::Error;

 /// Backend outputing events to a standalone HTML document containing a timetable
 pub struct HTMLBackend {}

 /// Options for the HTML backend
+#[derive(Default)]
 pub struct HTMLBackendOptions {
     /// Path to the template file. If not set, the default template (`data/template.html`) will be used.
     pub template_path: Option<String>,
 }

-/// Options of the HTML backend
-static OPTIONS: Mutex<HTMLBackendOptions> = Mutex::new(HTMLBackendOptions {
-    template_path: None,
-});
+impl HTMLBackendOptions {
+    fn get_template(&self) -> Result<String, std::io::Error> {
+        match self.template_path.as_ref() {
+            None => Ok(include_str!("../../data/template.html").to_string()),
+            Some(path) => std::fs::read_to_string(path),
+        }
+    }
+}

 /// Error that can occur during the compilation of the HTML backend
 #[derive(Debug, Error)]
@@ -29,14 +34,6 @@ pub enum HTMLBackendCompilationError {
     CouldNotReadTemplate(#[from] std::io::Error),
 }

-fn get_template() -> Result<String, std::io::Error> {
-    let options = OPTIONS.lock().unwrap();
-    match options.template_path.as_ref() {
-        None => Ok(include_str!("../../data/template.html").to_string()),
-        Some(path) => std::fs::read_to_string(path),
-    }
-}
-
 fn event_to_string(event: &Event) -> String {
     let duration = std::cmp::max(event.duration / 30, 1);
     let class = event.event_type.to_string();
@@ -51,19 +48,19 @@ fn event_to_string(event: &Event) -> String {
     res
 }

-impl HTMLBackend {
-    /// Configure the HTML backend
-    pub fn configure(opts: HTMLBackendOptions) {
-        *OPTIONS.lock().unwrap() = opts;
-    }
-}
-
-impl CompilingPass<Vec<Event>> for HTMLBackend {
+impl CompilingPass<Vec<Event>, HTMLBackendOptions> for HTMLBackend {
     type Residual = String;
     type Error = HTMLBackendCompilationError;

     fn apply(events: Vec<Event>) -> Result<Self::Residual, Self::Error> {
-        let template = get_template()?;
+        Self::apply_with(events, HTMLBackendOptions::default())
+    }
+
+    fn apply_with(
+        events: Vec<Event>,
+        options: HTMLBackendOptions,
+    ) -> Result<Self::Residual, Self::Error> {
+        let template = options.get_template()?;
         let mut str = String::new();
         str.push_str("<table><tbody>\n");
         for i in 9..21 {

hreymond committed:
Added ToHTML trait

hreymond committed:
Updated todos

hreymond committed:
Use contextual passes to provide parameters to the HTML backend

Removed the old shared variable "OPTIONS" and moved to a contextual pass.

hugnata commented on src/main.rs:
```diff
@@ -51,9 +86,39 @@ fn generate_tikz(content: &str) -> Result<String, TikzBackendCompilationError> {
+        .chain_pass_with::<HTMLBackend, HTMLBackendOptions>(options)
+}
+
+/// Write the output to a file or to stdout
```
I moved the "writing to the output file" logic out of the mess that was l.140.

After a bit of refactoring, I'm not sure it is necessary to have a separated function. Any thoughs on this ?

hreymond commented on src/main.rs:
```diff
@@ -51,9 +86,39 @@ fn generate_tikz(content: &str) -> Result<String, TikzBackendCompilationError> {
+        .chain_pass::<ParseTimetable>()?
+        .chain_pass_with::<HTMLBackend, HTMLBackendOptions>(options)
+}
+
```
I moved the "writing to the output file" logic out of the mess that was l.140.

After a bit of refactoring, I'm not sure it is necessary to have a separated function. Any thoughs on this ?

Lymkwi replied:
We're at such a level of overengineering, I don't see how factoring out logic like that could hurt.

hreymond commented on src/passes/html.rs:
```diff
@@ -0,0 +1,103 @@
+    #[error("Error while trying to read the template file: {0}")]
+    CouldNotReadTemplate(#[from] std::io::Error),
+}
+
```
I removed Clippy warnings about using `template_path.map_or_else()` instead of the match structure, since I find it easier to read

Lymkwi replied:
Agreed. It saves in performance to not do it that way but who cares.

hreymond commented on src/passes.rs:
```diff
@@ -1,11 +1,42 @@
 pub mod latex;
 pub mod parser;
 
 /// A trait defining compilation passes
```
I added documentation for the `CompilingPass` trait + a simple example.
I don't know the standard for documenting Rust, and I'm afraid to say wrong things.
Can you check it out ?

hreymond reviewed (commented):
I added the contextual pass management proposed by @Lymkwi, so now the parameters are given to the HTML backend with ` .chain_pass_with::<HTMLBackend, HTMLBackendOptions>(options)`
I used clippy to fix few things, and removed clippy warnings for other parts.

hugnata commented:
Is there a way to change the status of the PR to "change not approved yet ?" (since the code changed a lot since @Lymkwi approved my changes).

Lymkwi commented:
> Is there a way to change the status of the PR to "change not approved yet ?" (since the code changed a lot since @Lymkwi approved my changes).

Someone can request another review from me on the Desktop UI.

hugnata commented:
You can merge when you want 😉
